[A1] Answered on `d0cb463d`.

Thank you for the blocker: the repro is exact and it is the right call. Gate 23d accepted a route that proves nothing, and it did so in the one place where being wrong is worst. Fixed, plus every follow-up in the "not blockers" table, plus two of the three scope facts. One file still, still `sw/builder/test_builder.py` only, still no import of LiteX or migen.

## Blocker: the dedicated-keyword route now checks the receiving signature

`_declared_params()` returns the parameters a `def` declares BY NAME, with `*args` and `**kwargs` deliberately excluded, and that exclusion is the whole point. A dedicated keyword counts as a proven handoff only when `MilanSoC.__init__` declares it; anything `**kwargs` would swallow fails closed. `optional_blocks` itself is checked the same way, since renaming that parameter feeds the entire tier-1 dict to `SoCCore` for the same one warning.

Your repro, verbatim (move `maap` out of the dict onto a keyword nothing declares):

| | gate 23d |
|---|---|
| before | GREEN. `... CONSUMED by main() (5 through optional_blocks, dedicated ['maap', 'render_lpf']) == ... silent-no-op class, closed)`, rc=0, `ALL GATES PASS` |
| after | RED. `AssertionError: maap: main() hands --no-maap to MilanSoC as a dedicated maap= keyword, but MilanSoC.__init__ declares no maap parameter - its trailing **kwargs swallows the value and forwards it to LiteX SoCCore, which logs ONE warning, ignores it, and builds a bitstream with the block still in it. Put maap in the optional_blocks dict, or declare the parameter` |

The matching negative control is in `test_optional_block_consumption_gate_bites`: it rewrites EVERY dict row in turn into an undeclared dedicated keyword and requires red naming the block. It skips any block that IS a declared parameter, so the control stays honest if the legitimate route ever grows a second member.

```
[gate 23d mutation] 6/6 dedicated keywords MilanSoC.__init__ never declares rejected by
                    block name (LiteX warns and ignores; the bitstream ships unpruned)
```

## The six follow-ups: all closed here, none deferred

You were right that the printed `==` chain was ahead of the evidence. Rather than trim the sentence I traced the rest of the chain, because the argument for doing it in 23d is that #144 skips in CI: gate 23f needs migen and LiteX, so on the `pyyaml`-only docs job NONE of these six were caught by anything, including the one that prunes a default build. `_optional_block_prune_chain()` walks the same AST already in hand across all four source hops.

The tests are dependency questions, not spellings: `_reads(node, name)` asks whether the expression still READS the parameter, so `render_lpf`, `bool(render_lpf)` and `bool(render_lpf) and x` all pass and `True` does not. Each message names the affected blocks.

| mutation in `sw/litex/milan_soc.py` | before | after: actual `AssertionError` |
|---|---|---|
| drop `optional_blocks=optional_blocks` at `:6144` (hop 2) | GREEN | RED. `MilanSoC.__init__: the MilanNIC(...) call passes no optional_blocks= keyword at all, which no longer reads its own optional_blocks parameter - the prune is dropped one hop short of the RTL for: datapath_probes, i2s_playback, latency_taps, maap, media_clock_servo, rx_mac_filter` |
| drop `optional_blocks=optional_blocks` at `:538` (hop 3) | GREEN | RED. `MilanNIC.__init__: the add_milan_datapath(...) call passes no optional_blocks= keyword at all, ... for: datapath_probes, i2s_playback, latency_taps, maap, media_clock_servo, rx_mac_filter` |
| `render_lpf=bool(render_lpf)` to `render_lpf=True` at `:6143` | GREEN | RED. `MilanSoC.__init__: the MilanNIC(...) call passes render_lpf=True, which no longer reads its own render_lpf parameter - the prune is dropped one hop short of the RTL for: render_lpf` |
| `blocks[k] = True` at `:1000` | GREEN | RED. `add_milan_datapath: blocks[k] = True no longer reads the caller's value, so EVERY dict-routed prune is a no-op: datapath_probes, i2s_playback, latency_taps, maap, media_clock_servo, rx_mac_filter` |
| `blocks["render_lpf"] = True` at `:995` | GREEN | RED. `render_lpf: add_milan_datapath sets blocks['render_lpf'] = True, which never reads its render_lpf parameter - the dedicated keyword is inert and the block ships PRESENT whatever the CLI asked for` |
| **`if not blocks[k]:` inverted at `:1002`** | GREEN | RED. `add_milan_datapath emits the prune parameter under if blocks[k]: and not if not blocks[k]: - this single line carries the polarity of the whole tier-1 table, and inverted it makes an UNFLAGGED build prune every one of them: datapath_probes, i2s_playback, latency_taps, maap, media_clock_servo, render_lpf, rx_mac_filter` |

Three more of the same family, found while writing the above and closed with it:

| mutation | before | after |
|---|---|---|
| `if not blocks[k]:` to `if False:` (no build ever prunes) | GREEN | RED, same message with `if False:` |
| `dict.fromkeys(MILAN_OPTIONAL_BLOCKS, True)` to `False` (default build prunes all seven) | GREEN | RED. `add_milan_datapath: blocks = dict.fromkeys(MILAN_OPTIONAL_BLOCKS, False) does not seed every block PRESENT from MILAN_OPTIONAL_BLOCKS - anything else prunes on an UNFLAGGED build: ...` |
| the emit guard grows an `else` arm | GREEN | RED, named |

## The three scope facts: two closed, one out of scope and why

| fact | before | after |
|---|---|---|
| a brand-new decorative `--no-widget` consumed by nothing | GREEN | RED. `--no-widget: main() declares it and never reads args.no_widget, so it is decorative - the argv records the intent and the gateware never hears about it (the #130 shape, one hop earlier)` |
| `default=True` on a `--no-*` flag | GREEN | RED. `--no-maap: default=True inverts the flag - an UNFLAGGED build would take the pruned path and ship the block missing, which is the silent no-op with its sign flipped` |
| a second `KL_maap` instantiated outside the `generate if (MAAP_P != 0)` arm | GREEN | still GREEN, deliberately |

The first two are now checked for EVERY `--no-*` flag `main()` declares, not only the seven in the table, so `--no-milan` and `--no-rx-rsc` are covered too. An explicit `dest=` is honoured.

The third I am leaving open on purpose rather than half-closing it. Doing it generically needs two things this PR should not invent: a MODULE-NAME column in `OPTIONAL_BLOCKS` (the `why` prose is not a parseable module list, two rows read `rx_mac_filter + tcam` and `the APRB + PBK probe groups (0x8B4-0x8D0)`), and a comment-stripping SV instantiation scanner to tell `KL_maap` in an instantiation from `KL_maap` in the eleven `//!` lines that mention it. A text gate that guessed at either would be brittle noise. The honest closer is an elaborated-hierarchy diff, pruned versus unpruned, out of the Verilator or Yosys flow, one level below where #144 watches. Happy to raise that as its own ticket.

## What is still #144's job, explicitly

Gate 23d now proves the SOURCE chain. It cannot prove what elaboration does, and I am not claiming it does. #144 (gate 23f) remains the only thing that grades:

- that the `dp_params` dict as built is the dict `Instance("milan_datapath", ...)` is actually handed. A later rebind, filter or shadow of `dp_params` is invisible to reading;
- that `--no-<block>` MOVES NOTHING ELSE, i.e. no cross-talk between blocks at elaboration;
- `sound_card`, which is graded upside down and is not in `OPTIONAL_BLOCKS` at all;
- anything migen or LiteX does rather than anything this file says.

The printed sentence now says so instead of ending on an unqualified "closed":

```
[gate 23d] 7 optional blocks: builder key == milan_soc.py map == a real off-by-default
  --no-* store_true main() READS == consumed by main() (6 through optional_blocks,
  dedicated ['render_lpf'], each a NAMED MilanSoC.__init__ parameter, none swallowed by
  **kwargs) == forwarded through MilanNIC and add_milan_datapath == emitted as
  `p_<PARAM> = 0` under `if not blocks[k]` == a milan_datapath `parameter int X = 1`
  guarding a generate WITH an else arm == the AREA_BUDGET tier-1 table row (the
  p_AAF_PLAYBACK silent-no-op class, closed on the source chain; what elaboration hands
  Instance() is gate 23f, task #144)
```

Heads-up for #144, which stacks on this branch: its PR body says the five edits it targets "leave every existing gate green". After this commit four of them are red in 23d as well. That is the two gates overlapping in the intended direction, not a conflict, but the sentence wants a rebase-time edit and #144's own controls call its grader directly, so they are unaffected.

## Your eight, all still red, all still by name

Re-run against `d0cb463d`:

| # | mutation | verdict |
|---|---|---|
| M1 | dict row deleted (the #130 shape) | RED, `maap` |
| M2 | dedicated `render_lpf=` keyword deleted | RED, `render_lpf` |
| M3 | polarity flip `"maap": args.no_maap` | RED, `maap` |
| M4 | cross-wire to `not args.no_latency_taps` | RED, `maap` |
| M5 | hard-code `"maap": True` | RED, `maap` |
| M6 | `MAAP_P` renamed to `MAAP_PP` | RED, `maap: builder says MAAP_P, milan_soc.py says 'MAAP_PP'` |
| M7 | `--no-maap` dropped from `emit_board_opts` argv | 23d green, suite RED at 23b: `MAAP pruned, static DMAC base: --no-maap missing from argv`, unchanged |
| M8 | `dp_params[f"p_{param}"] = 1` | RED, now from the chain too: `only 0 prunes, any other value leaves the generate arm elaborated (AREA_BUDGET rule 1)` |

## Verification

```
$ git status --porcelain                          (empty, after every mutation and at the end)
$ python3 -m pytest -q sw/builder/test_builder.py
54 passed in 9.13s                                (before: 54 passed in 8.43s)
$ python3 sw/builder/test_builder.py ; echo rc=$?
  [gate 23d] ...
  [gate 23d mutation] 7/7 disconnected --no-* handoffs rejected by block name
  [gate 23d mutation] 6/6 dedicated keywords MilanSoC.__init__ never declares rejected by block name
  [gate 23d chain] 10/10 cut links between main() and `dp_params[f"p_{param}"] = 0` rejected,
                   the three that would prune an UNFLAGGED build included
ALL GATES PASS
rc=0
$ python3 scripts/lint_rtl.py --check
LINT GATE: PASS (97 violation(s) <= ratchet 99; 22 waived, 0 justified lint_off)
$ python3 scripts/docs_check.py
docs_check: 0 finding(s) across 211 md files [git ls-files]
$ git diff --name-only origin/dev...HEAD
sw/builder/test_builder.py
$ git diff --check origin/dev...HEAD               (no output, rc=0)
$ python3 -c 'import litex' ; python3 -c 'import migen'
ModuleNotFoundError                                (both, on the box every number above came from)
```

Cost: the two functions are 86 ms and 894 ms, against 76 ms and 259 ms before. The whole of the 0.6 s is 23 fresh `ast.parse` calls over a 7000-line file, one per negative control, and it is unavoidable because each control mutates the text. Suite 8.4 s to 9.1 s. Three consecutive runs: rc=0 each, and the `[gate 23d]`, `[gate 23e]`, both `[gate 23d mutation]` and `[gate 23d chain]` lines are byte-identical across all three.

No em dash on any added line. No RTL and no `sw/litex/milan_soc.py` change: every mutation above was applied on disk, graded, and reverted, with `git status` verified clean after each.

--- a/sw/builder/test_builder.py
+++ b/sw/builder/test_builder.py
@@ -3686,8 +3686,13 @@ def test_optional_block_names_reach_the_rtl():
           "`p_<PARAM> = 0` under `if not blocks[k]` == a milan_datapath "
           "`parameter int X = 1` guarding a generate WITH an else arm == the "
           "AREA_BUDGET tier-1 table row (the p_AAF_PLAYBACK silent-no-op "
-          "class, closed on the source chain; what elaboration hands "
-          "Instance() is gate 23f, task #144)")
+          "class, closed for the SPELLING of these statements. A statement "
+          "that rebinds or overrides a carrier without touching them - "
+          "reassigning optional_blocks, overriding a fold, deleting a "
+          "dp_params key - is invisible here by construction, because each "
+          "hop is checked in isolation and _reads() is a name-appearance "
+          "test. That family and what elaboration hands Instance() are both "
+          "gate 23f, task #144)")
 
     # gate 23e (task #65): the loopback lane rides the SAME decorative-ABI
     # risk but the OPPOSITE polarity - it is an ADD, default OFF, so it is

